swap_server_client.go

```diff
@@ -202,16 +209,62 @@ func (s *grpcSwapServerClient) GetLoopInQuote(ctx context.Context,
 
 	rpcCtx, rpcCancel := context.WithTimeout(ctx, globalCallTimeout)
 	defer rpcCancel()
-	quoteResp, err := s.server.LoopInQuote(rpcCtx,
-		&looprpc.ServerLoopInQuoteRequest{
-			Amt:             uint64(amt),
-			ProtocolVersion: loopdb.CurrentRPCProtocolVersion,
-		},
-	)
+
+	quoteReq := &looprpc.ServerLoopInQuoteRequest{
+		Amt:             uint64(amt),
+		ProtocolVersion: loopdb.CurrentRPCProtocolVersion,
+	}
+
+	var probeHash lntypes.Hash
+	if loopdb.CurrentRPCProtocolVersion == looprpc.ProtocolVersion_MULTI_LOOP_IN_V2 {
+		// Generate random preimage.
+		var preimage lntypes.Preimage
+		if _, err := rand.Read(preimage[:]); err != nil {
+			log.Error("Cannot generate preimage")
+		}
+
+		probeHash := lntypes.Hash(sha256.Sum256(preimage[:]))
+
+		log.Infof("Creating quote probe invoice %v", probeHash)
+		probeInvoice, err := s.lnd.Invoices.AddHoldInvoice(
+			rpcCtx, &invoicesrpc.AddInvoiceData{
+				Hash:   &probeHash,
+				Value:  lnwire.NewMSatFromSatoshis(amt),
+				Memo:   "loop in quote probe",
+				Expiry: 3600,
+			},
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		quoteReq.ProbeInvoice = probeInvoice
+	}
+
+	quoteResp, err := s.server.LoopInQuote(rpcCtx, quoteReq)
 	if err != nil {
 		return nil, err
 	}
 
+	if loopdb.CurrentRPCProtocolVersion == looprpc.ProtocolVersion_MULTI_LOOP_IN_V2 {
+		// Create a cancellable context that is used for monitoring the probe.
+		probeWaitCtx, probeWaitCancel := context.WithCancel(ctx)
```

Comment:
Don't we need to create this with a timeout so that it'll cancel after some amount of time?